Add Blogbost: Advanced Flink Application Patterns: Custom Window Proc… (#362)
apache/flink-web: _posts/2020-07-23-demo-fraud-detection-3.md

---
layout: post
title: "Advanced Flink Application Patterns Vol.3: Custom Window Processing"

> date: 2020-07-23T12:00:00.000Z

Reviewer: ```suggestion
date: 2020-07-27T12:00:00.000Z
```

partitioning key.

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-3/time-windows.png" width="600px" alt="Figure 1: Time Windows"/>
<br/>
<i><small>Figure 1: Time Windows</small></i>
</center>
<br/>

One of the common key requirements for a fraud detection system is *low
response time*. The sooner the fraudulent action gets detected, the
higher the chances that it can be blocked and its negative consequences
mitigated. This requirement is especially prominent in the financial
domain, where you have one important constraint - any time spent
evaluating a fraud detection model is time that a law-abiding user of
your system will spend waiting for a response. Swiftness of processing
often becomes a competitive advantage between various payment systems
and the time limit for producing an alert could lie as low as *300-500
ms*. This is all the time you get from the moment of ingestion of a
transaction event into a fraud detection system until an alert has to
become available to downstream systems. 

As you might know, Flink provides a powerful [Window
API](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/operators/windows.html)

> that is applicable for a wide range of use cases. However, If you go

Reviewer: ```suggestion
that is applicable for a wide range of use cases. However, if you go
```

over all of the available types of supported windows, you will realize
that none of them exactly match our main requirement for this use case -
the low-latency evaluation of *each* incoming transaction. There is
no type of window in Flink that can express the *"x minutes/hours/days
back from the <u>current event</u>"* semantic. In the Window API, events
fall into windows (as defined by the window

> [assigners](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/operators/windows.html#window-assigners)),

Reviewer: it is probably better to link to flink-docs-release-1.11 so that the link does not become outdated if 1.12 changes structure.
(applies to all links in the post)

Author: Good idea, changed the links.

but they cannot themselves individually control the creation and
evaluation of windows\*. As described above, our goal for the fraud
detection engine is to achieve immediate evaluation of the previous
relevant data points as soon as the new event is received. This raises

> the question of feasibility of applying the Window API in this case.

Reviewer: Not sure, but should we at least mention custom window assigners,... that could lead to a similar solution but are more complex to handle?

Author: I've looked into this approach a while ago and came to the conclusion that it leads to some overly complex implementation, not even taking into account that it won't work with the dynamic rules updates - windows have to be statically defined. Not sure how much value it adds to mention something we do not recommend and that does not match the use case requirements (broadcasted rules).

Reviewer: I guess, that is exactly what I proposed to add (that there is something but that's not really feasible).

Something around the line of
> The window API offers some customization options for defining custom triggers, evictors, and window assigners which may get to a similar result. However, it is usually difficult to get this right (and easy to break) while at the same time, business logic based on a `ProcessFunction` is easier to understand and maintain and offers access to broadcast state as well.

Author: Thanks, @NicoK , I think your proposition sounds very good. I am adding it.

might be different events with exactly the same timestamp, therefore
instead of individual Transaction per key(timestamp) we will store sets.

```java
MapState<Long, Set<Transaction>> windowState;
```

<div class="alert alert-info" markdown="1">
<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Side Note </span>
when any Flink-managed state is used inside a
`KeyedProcessFunction`, the data returned by the `state.value()` call is
automatically scoped by the key of the *currently-processed event*
- see Figure 4. If `MapState` is used, the same principle applies, with
the difference that a `Map` is returned instead of `MyObject`. If you are
compelled to do something like
`mapState.value().get(inputEvent.getKey())`, you should probably be using
`ValueState` instead of the `MapState`. As we want to store *multiple values
per event key*, in our case, `MapState` is the right choice.
</div>

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-3/keyed-state-scoping.png" width="800px" alt="Figure 4: Keyed State Scoping"/>
<br/>
<i><small>Figure 4: Keyed State Scoping</small></i>
</center>

> <br/>

Reviewer: nice excursion!

Do you think, this figure should rather be inside the info box?

> As described in the [first]({{ site.baseurl }}/news/2020/01/15/demo-fraud-detection.html) blog of the series, we are dispatching events based on the keys

Reviewer: ```suggestion
As described in the [first blog of the series]({{ site.baseurl }}/news/2020/01/15/demo-fraud-detection.html), we are dispatching events based on the keys
```

for overlapping time windows, and therefore unnecessarily store
duplicate events. A better approach is to always store just enough data
to be able to estimate all currently active rules which are scoped by
the same key. In order to achieve that, whenever a new rule is added, we
will determine if its time window has the largest span and store it in
the broadcast state under the special reserved `WIDEST_RULE_KEY`. This
information will later be used during the state clean-up procedure, as
described later in this section.

```java
@Override
public void processBroadcastElement(Rule rule, Context ctx, Collector<Alert> out){
  ...
  updateWidestWindowRule(rule, broadcastState);
}

private void updateWidestWindowRule(Rule rule, BroadcastState<Integer, Rule> broadcastState){
  Rule widestWindowRule = broadcastState.get(WIDEST_RULE_KEY);

  if (widestWindowRule == null) {
    broadcastState.put(WIDEST_RULE_KEY, rule);
    return;
  }

  if (widestWindowRule.getWindowMillis() < rule.getWindowMillis()) {
    broadcastState.put(WIDEST_RULE_KEY, rule);
  }
}
```

Let's now look at the implementation of the main method,
`processElement()`, in some detail.

> In the previous [blog post]({{ site.baseurl }}/news/2020/01/15/demo-fraud-detection.html#dynamic-data-partitioning), we described how `DynamicKeyFunction` allowed

Reviewer: ```suggestion
In the [previous blog post]({{ site.baseurl }}/news/2020/01/15/demo-fraud-detection.html#dynamic-data-partitioning), we described how `DynamicKeyFunction` allowed
```

us to perform dynamic data partitioning based on the `groupingKeyNames`
parameter in the rule definition. The subsequent description is focused
around the `DynamicAlertFunction`, which makes use of the remaining rule
settings.

<center>
<img src="{{ site.baseurl }}/img/blog/patterns-blog-3/sample-rule-definition.png" width="700px" alt="Figure 5: Sample Rule Definition"/>
<br/>
<i><small>Figure 5: Sample Rule Definition</small></i>
</center>
<br/>

As described in the previous parts of the blog post
series, our alerting process function receives events of type
`Keyed<Transaction, String, Integer>`, where `Transaction` is the main
"wrapped" event, String is the key (*payer \#x - beneficiary \#y* in
Figure 1), and `Integer` is the ID of the rule that caused the dispatch of

> this event. This rule was previously [stored]({{ site.baseurl }}/news/2020/03/24/demo-fraud-detection-2.html#broadcast-state-pattern) in the broadcast state and has to be retrieved from that state by the ID. Here is the

Reviewer: ```suggestion
this event. This rule was previously [stored in the broadcast state]({{ site.baseurl }}/news/2020/03/24/demo-fraud-detection-2.html#broadcast-state-pattern) and has to be retrieved from that state by the ID. Here is the
```

    Long windowStartTimestampForEvent = rule.getWindowStartTimestampFor(currentEventTime);
    SimpleAccumulator<BigDecimal> aggregator = RuleHelper.getAggregator(rule);

    for (Long stateEventTime : windowState.keys()) {
      if (isStateValueInWindow(stateEventTime, windowStartForEvent, currentEventTime)) {
        aggregateValuesInState(stateEventTime, aggregator, rule);
      }
    }

    // Evaluate the rule and trigger an alert if violated
    BigDecimal aggregateResult = aggregator.getLocalValue();
    boolean isRuleViolated = rule.apply(aggregateResult);
    if (isRuleViolated) {
      long decisionTime = System.currentTimeMillis();
      out.collect(new Alert<>(rule.getRuleId(),
                              rule,
                              value.getKey(),
                              decisionTime,
                              value.getWrapped(),
                              aggregateResult));
    }

    // Register timers to ensure state cleanup
    long cleanupTime = (currentEventTime / 1000) * 1000;
    ctx.timerService().registerEventTimeTimer(cleanupTime);
  }
```

<br/>
Here are the details of the steps:  

> 1)  We first add each new event to our window state:

Reviewer: Can you add these numbers into the code above, e.g. like this? That way, people can better see what it relates to.

For example by marking lines with `// <--- #...` or like this (or maybe you have some other idea):
```
// 1) Add Transaction to state
...
// Calculate the aggregate value
// 2) retrieve broadcased rule
```

  Set<Transaction> inWindow = windowState.get(stateEventTime);
  for (Transaction event : inWindow) {
    BigDecimal aggregatedValue =
        FieldsExtractor.getBigDecimalByName(rule.getAggregateFieldName(), event);
    aggregator.add(aggregatedValue);
  }
}
```

5) Having an aggregate value, we can compare it to the threshold value
    that is specified in the rule definition and fire an alert, if
    necessary.

6) At the end, we register a clean-up timer using
`ctx.timerService().registerEventTimeTimer()`. This timer will be
    responsible for removing the current transaction when it is going to
    move out of scope.

<div class="alert alert-info" markdown="1">
<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span>  Note </span>
Notice the rounding during timer creation. It is an important technique
which enables a reasonable trade-off between the precision with which
the timers will be triggered, and the number of timers being used.
Timers are stored in Flink's fault-tolerant state, and managing them
with millisecond-level precision can be wasteful. In our case, with this

> rounding, we will create at most one timer per key in any given second.

Reviewer: You could link to https://ci.apache.org/projects/flink/flink-docs-release-1.11/dev/stream/operators/process_function.html#timer-coalescing if you want.

      if (stateEventTime < threshold) {
        keys.remove();
      }
    }
  } catch (Exception ex) {
    throw new RuntimeException(ex);
  }
}
```

<div class="alert alert-info" markdown="1">
<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span>  Note</span>
You might be wondering why we did not use `ListState` , as we are always
iterating over all of the values of the window state? This is actually
an optimization for the case when `RocksDBStateBackend`
[is used](https://ci.apache.org/projects/flink/flink-docs-stable/ops/state/state_backends.html#the-rocksdbstatebackend). Iterating over a `ListState` would cause all of the `Transaction`
objects to be deserialized. Using `MapState`\'s keys iterator only causes
deserialization of the keys (type `long`), and therefore reduces the
computational overhead.
</div>

This concludes the description of the implementation details. Our
approach triggers evaluation of a time window as soon as a new
transaction arrives. It therefore fulfills the main requirement that we
have targeted - low delay for potentially issuing an alert. For the

> complete implementation, please follow
> [this](https://github.com/afedulov/fraud-detection-demo) link.

Reviewer: ```suggestion
complete implementation, please have a look at
[the project on github](https://github.com/afedulov/fraud-detection-demo).
```

and verify where your application might be further optimized by ensuring
that [more efficient
serializers](https://flink.apache.org/news/2020/04/15/flink-serialization-tuning-vol-1.html#performance-comparison)
are being used.

<div class="alert alert-info" markdown="1">
<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span>  Tip:</span>
you can quickly determine which serializer is going to be
used for your classes by setting a breakpoint and verifying the type of
the returned TypeInformation.
</div>

<center>
<table class="tg">
  <tr>
    <td class="tg-topcenter">
      <img src="{{ site.baseurl }}/img/blog/patterns-blog-3/type-pojo.png" alt="POJO"/></td>
    <td class="tg-top">
      <i>PojoTypeInfo</i> indicates that that an efficient Flink POJO serializer will be used.</td>
  </tr>
  <tr>
    <td class="tg-top">
      <img src="{{ site.baseurl }}/img/blog/patterns-blog-3/type-kryo.png" alt="Kryo"/></td>
    <td class="tg-topcenter">
      <i>GenericTypeInfo</i> indicates the fallback to a Kryo serializer.</td>
  </tr>

> </table>

Reviewer: Can you put this table inside the info box as well?
